app/jobs/populate_fake_data_job.rb

```diff
@@ -0,0 +1,285 @@
+            :appointment,
+            patient: args[:patient],
+            user: user,
+            creation_facility: user.facility,
+            facility: user.facility
+          ))
+        },
+        patient_sample_size: 0.50,
+        request_key: :appointments,
+        api_version: 'v3'
+      },
+
+      overdue_appointment: {
+        size_fn: -> { rand(1..2) },
+        build_fn: -> (args) {
+          return if args[:patient].latest_scheduled_appointment.present?
+
+          build_appointment_payload(FactoryBot.build(
+            :appointment,
+            :overdue,
+            patient: args[:patient],
+            user: user,
+            creation_facility: user.facility,
+            facility: user.facility
+          ))
+        },
+        patient_sample_size: 0.50,
+        request_key: :appointments,
+        api_version: 'v3'
+      },
+
+      completed_phone_call: {
+        time_fn: -> { Faker::Time.between(from: 9.months.ago, to: Date.today) },
+        size_fn: -> { rand(1..1) },
+        build_fn: -> (args) {
+
+          FactoryBot.create(
+            :call_log,
+            result: 'completed',
+            caller_phone_number: user.phone_number,
+            callee_phone_number: args[:patient].latest_phone_number,
+            end_time: args[:time_fn].call
+          )
+        },
+        patient_sample_size: 0.20
+      }
+    }
+  end
+
+  def perform(user_id)
+    return if ENV['SIMPLE_SERVER_ENV'] == 'production'
+
+    @user = User.find(user_id)
+
+    #
+    # register some patients with their medical histories
+    #
+    patient_trait_args = patient_traits[:registered_patient]
+    number_of_patient_records = patient_trait_args[:size_fn].call.to_i
+    registered_patients = number_of_patient_records.times.flat_map do
+      generate(patient_trait_args)
+    end
+    create_resources(registered_patients, :registered_patient, patient_trait_args)
+
+    diagnosis_trait_args = patient_traits[:diagnosis]
+    diagnosis_data = user.registered_patients.flat_map do |patient|
+      generate(diagnosis_trait_args.merge(patient: patient))
+    end
+    create_resources(diagnosis_data, :diagnosis, diagnosis_trait_args)
+
+    #
+    # generate various traits for the registered patients
+    #
+    traits.each do |trait_name, trait_args|
+      create_resources(generate_traits(trait_args), trait_name, trait_args)
+    end
+  end
+
+  private
+
+  def create_resources(data, trait_name, trait_args)
+    puts "Creating #{trait_name} for #{user.full_name}..."
+
+    request_key = trait_args[:request_key]
+    api_version = trait_args[:api_version]
+    return if request_key.blank?
+
+    logger.info("Creating #{trait_name} for #{user.full_name} " +
+                  "with #{data.size} #{request_key} – facility: #{user.facility.name}")
+
+    data.each_slice(SYNC_PAYLOAD_SIZE) do |data_slice|
+      api_post("/api/#{api_version}/#{request_key}/sync", request_key => data_slice) if data_slice.present?
```

Comment:
Can you say more about why we are going through our API to generate seed data? Are there service objects or other business constraints enforced in the API (I haven't really looked at our API yet, so I'm ignorant here =)) 

Our model layer _should_ have all the data consistency / validations enforced, so using the models directly would be fine for this I think. My preference for seed generation would be to keep it as simple as possible via FactoryBot calls or plain old `Model.create` calls, so we can change it and maintain it easily over time.

Reply:
Yeah, ideally our data layer should not be different from what you'd do through the API. That's however, currently not the case. 

We do a few things that are special in APIs:

1. Handling some tricky edge-cases around API compatibility across versions and sanitizing data accordingly in the pre-Model layers (Transformers, Validators)
2. Creating Encounters behind the scenes as and when BloodPressures or BloodSugars are synced. This could be done by simply creating an Encounter factory in here and linking the children over, but given 1. and our past experiences with datagen not going so well because stuff is happening behind the scenes, I'd chosen an ostensibly less worrisome path albeit a costlier one (in terms of speed and potentially maintenance). Though I'd argue it keeps us closer to how a real user would generate data, as it stands.

Reply:
> * Handling some tricky edge-cases around API compatibility across versions and sanitizing data accordingly in the pre-Model layers (Transformers, Validators)

This wouldn't impact seed data generation, right?

> * Creating Encounters behind the scenes as and when BloodPressures or BloodSugars are synced. This could be done by simply creating an Encounter factory in here and linking the children over, but given 1. and our past experiences with datagen not going so well because stuff is happening behind the scenes, I'd chosen an ostensibly less worrisome path albeit a costlier one (in terms of speed and potentially maintenance). Though I'd argue it keeps us closer to how a real user would generate data, as it stands.

Can you point me to the relevant Encounter code in the API? Or maybe give me a code walkthrough tomorrow ? 😄 I'm pretty free all morning eastern US time. 

It sounds like getting Encounter syncing into a proper service object decoupled from the API (possibly fired via a callback) is a pretty high priority to make our data model easier to work with and test.

I totally agree we want seed data to be correct and in the same state as when users sync it. My main point is that if we have to use the API to do it, something is not good and we gotta unravel that before it gets worse.

Reply:
@rsanheim you're absolutely right. Our model layer doesn't represent all the constraints necessary for coherent data, so we're forced to use our API layer. As Kitty mentions below, I don't think this needs to block this PR in particular, as fixing this model issue is a larger effort. Let's walk through it together soon!